examples/mpg_example.py

```diff
@@ -0,0 +1,308 @@
+  """Create a DatasetMetadata for the raw data."""
+  column_schemas = {
+    key: dataset_schema.ColumnSchema(
+       tf.string, [], dataset_schema.FixedColumnRepresentation())
+    for key in CATEGORICAL_FEATURE_KEYS
+  }
+  column_schemas.update({
+    key: dataset_schema.ColumnSchema(
+       tf.float32, [], dataset_schema.FixedColumnRepresentation())
+    for key in NUMERIC_FEATURE_KEYS
+  })
+  column_schemas[LABEL_KEY] = dataset_schema.ColumnSchema(
+     tf.float32, [], dataset_schema.FixedColumnRepresentation())
+  raw_data_metadata = dataset_metadata.DatasetMetadata(dataset_schema.Schema(
+     column_schemas))
+  return raw_data_metadata
+
+
+RAW_DATA_METADATA = _create_raw_metadata()
+
+# Constants used for training.  Note that the number of instances will be
+# computed by tf.Transform in future versions, in which case it can be read from
+# the metadata.  Similarly BUCKET_SIZES will not be needed as this information
+# will be stored in the metadata for each of the columns.  The bucket size
+# includes all listed categories in the dataset description as well as one extra
+# for "?" which represents unknown.
+BATCH_SIZE = 5
+TRAIN_NUM_EPOCHS = 20
+NUM_TRAIN_INSTANCES = 340
+NUM_TEST_INSTANCES = 50
+BUCKET_SIZES = [5, 12, 1024, 3]
+
+EXPORTED_MODEL_DIR = 'exported_model_dir'
+
+
+def create_transform_fn(train_data_file, working_dir):
+  """Create a transform function that can be run on-the-fly while training
+
+  Read in the data using the CSV reader, and transform it using a
+  preprocessing pipeline that scales numeric data and converts categorical data
+  from strings to int64 values indices, by creating a vocabulary for each
+  category.
+
+  Args:
+    train_data_file: File containing training data
+    working_dir: Directory to write transformed data and metadata to
+  """
+
+  def preprocessing_fn(inputs):
+    """Preprocess input columns into transformed columns."""
+    outputs = {}
+
+    # Scale numeric columns to have range [0, 1].
+    for key in NUMERIC_FEATURE_KEYS:
+      outputs[key] = tft.scale_to_0_1(inputs[key])
+
+    # For all categorical columns except the label column, we use
+    # tft.string_to_int which computes the set of unique values and uses this
+    # to convert the strings to indices.
+    for key in CATEGORICAL_FEATURE_KEYS:
+      outputs[key] = tft.string_to_int(inputs[key])
+
+    # For the label column we provide the mapping from string to index.
+    outputs[LABEL_KEY] = inputs[LABEL_KEY]
+
+    return outputs
+
+  # The "with" block will create a pipeline, and run that pipeline at the exit
+  # of the block.
+  with beam.Pipeline() as pipeline:
+    with beam_impl.Context(temp_dir=tempfile.mkdtemp()):
+      # Create a coder to read the mpg data with the schema.  To do this we
+      # need to list all columns in order since the schema doesn't specify the
+      # order of columns in the csv.
+      converter = csv_coder.CsvCoder(ordered_columns, RAW_DATA_METADATA.schema)
```

Comment:
We are using a different csv_coder (tf.decode_csv) when parsing training samples (see line 172). Don't know if that might be an issue.

Reply:
My belief is that if a given example parses successfully with both coders, the results are very likely to be the same.  But in edge cases the coders may differ in what they consider a valid input.  Ultimately we want to harmonize the parsing done in tf.Transform with the in-graph parsing done with `tf.decode_csv` and `tf.parse_example`, but that is a work in progress.

For now I would say proceed assuming they are the same but let us know if any discrepancies emerge.

Reply:
I agree